### PR TITLE
restore hidden states added

### PR DESCRIPTION
**Pull request by hnekoeiq**

### CODE REVIEW

**Comment by hnekoeiq** [comment]
This PR adds two features to the DRQN agent:
1- Restoring hidden states from the replay buffer
2- Burn-in frames to warm up the rnn module.

File: hive/agents/drqn.py [hunk 1]
```diff
@@ -164,6 +182,46 @@ def create_q_networks(self, representation_net):
...
+                        "cell_state": self._prev_cell_state,
+                    }
+                )
+
+            elif self._rnn_type == "gru":
+                preprocessed_update_info.update(
+                    {
+                        "hidden_state": self._prev_hidden_state,
+                    }
+                )
```
Comment (kshitijkg) [inline hive/agents/drqn.py]: This works too, but it might be easier to adapt this code to other memory based architecture if we instead store "memory" that can contain any kind of memory (hidden, hidden+cell, trxl memory), etc and have a function that can pack and unpack this memory? So the user has to only modify teh pack unpack function when adding a new memory based architecture?

Comment (hnekoeiq) [reply hive/agents/drqn.py]: @kshitijkg With the most recent commit, the update and act functions are almost independent of the type of memory (I'm still calling it hidden_state because it is a DRQN agent, but we can replace everything with memory). Could you please review it again?

File: hive/agents/drqn.py [hunk 2]
```diff
@@ -277,7 +316,20 @@ def update(self, update_info, agent_traj_state=None):
...
+                interm_loss = self._loss_fn(pred_qvals, q_targets)
+                mask = torch.zeros(
+                    self._replay_buffer._max_seq_len,
+                    device=self._device,
+                    dtype=torch.float,
+                )
+                mask[self._burn_frames :] = 1.0
+                mask = mask.view(1, -1)
+                interm_loss *= mask
+                loss = interm_loss.mean()
```
Comment (mrsamsami) [inline hive/agents/drqn.py:380]: Isn't the correct way of doing it `interm_loss.sum() / mask.sum()`?

File: hive/agents/drqn.py [hunk 3]
```diff
@@ -293,3 +345,65 @@ def update(self, update_info, agent_traj_state=None):
...
+                f"received {self._rnn_type}."
+            )
+
+        return hidden_state
+
+    def get_hidden_state(self, batch):
+        if self._store_hidden == True:
+            hidden_state = (
+                torch.tensor(
+                    batch["hidden_state"][:, 0].squeeze(1).squeeze(1).unsqueeze(0),
```
Comment (mrsamsami) [inline hive/agents/drqn.py]: I'd suggest using `view()` or `reshape()` to potentially make the code cleaner.

File: hive/agents/drqn.py (hunk 2 above)
Comment (hnekoeiq) [reply hive/agents/drqn.py:380]: That's true. This should be fixed after merging #339.
